### PR TITLE
tests: unittests for multipath

### PR DESCRIPTION
**Pull request by raharper**
- Add unittests for multipath storage module.
- Add test baseclass CiTestCase for common helpers
- Add random_string class method on CiTestCase

### CODE REVIEW

**Comment by raharper** [comment]
Was looking to fix a recently filed bug, but that's fixed, so I added unittests to confirm master is good to go.

**Comment by xnox** [comment]
Looks good. I do wonder if there are any better ways of scrapping the data we need...

File: probert/multipath.py [hunk 1]
```diff
@@ -42,6 +43,9 @@ def _extract_mpath_data(cmd, show_verb):
     for line in data.splitlines():
         mp_dict = None
         try:
+            field_vals = line.split()
+            log.debug('Extracted multipath %s fields: %s',
+                      show_verb, field_vals)
             mp_dict = mptype(*line.split())._asdict()
```
Comment (OddBloke) [inline probert/multipath.py]: Should we do `*field_vals` here to avoid splitting twice?

File: probert/tests/helpers.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import random
+import string
+import testtools
+
+class CiTestCase(testtools.TestCase):
+
+    @classmethod
+    def random_string(cls, length=8):
```
Comment (OddBloke) [inline probert/tests/helpers.py]: We haven't had this conversation in the context of probert, but ***if*** we want to migrate to pytest in the future, we'll find it easier if helpers (that don't require `self`) aren't attached to test classes.

Could this be a function that `test_multipath.py` just imports directly?

File: probert/tests/test_multipath.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from probert import multipath
+from probert.tests.helpers import CiTestCase
+
+
+class TestMultipath(CiTestCase):
+
+    @mock.patch('probert.multipath.subprocess.run')
+    def test_multipath_show_paths(self, m_run):
+        mp_out = " ".join([self.random_string() for x in range(0, 8)])
+        cp = subprocess.CompletedProcess(args = ['foo'], returncode = 0,
```
Comment (OddBloke) [inline probert/tests/test_multipath.py]: Not super-important, but FYI the spacing around the `=`s for keyword args here is non-standard per [PEP-8](https://www.python.org/dev/peps/pep-0008/#other-recommendations):

> Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter.

File: probert/tests/test_multipath.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        m_run.return_value = cp
+        expected_result = [multipath.MPath(*mp_out.split())._asdict()]
+        result = multipath.multipath_show_paths()
+        self.assertEqual(expected_result, result)
+
+    @mock.patch('probert.multipath.subprocess.run')
+    def test_multipath_show_paths_skips_unparsable_output(self, m_run):
+        lines = ["1 2 3 4 5 6 7 8", "", "a b c d e f g h"]
+        mp_out = "\n".join(lines)
+        cp = subprocess.CompletedProcess(args = ['foo'], returncode = 0,
```
Comment (OddBloke) [inline probert/tests/test_multipath.py]: `subprocess.CompletedProcess` was introduced in 3.5; does probert only target Ubuntu systems that are xenial or later?

File: probert/tests/test_multipath.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                                         stdout = mp_out.encode('utf-8'),
+                                         stderr = "")
+        m_run.return_value = cp
+        expected_result = [multipath.MPath(*mp_out.split())._asdict()]
+        result = multipath.multipath_show_paths()
+        self.assertEqual(expected_result, result)
+
+    @mock.patch('probert.multipath.subprocess.run')
+    def test_multipath_show_paths_skips_unparsable_output(self, m_run):
```
Comment (OddBloke) [inline probert/tests/test_multipath.py]: The unparsable output here is just the empty string, IIRC. Should we add more lines with different garbage to make this test more robust, or are we intentionally only testing an empty line?

(If the latter, `test_multipath_show_paths_skips_blank_lines` might be a more apt name?)

File: probert/tests/test_multipath.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        cp = subprocess.CompletedProcess(args = ['foo'], returncode = 0,
+                                         stdout = mp_out.encode('utf-8'),
+                                         stderr = "")
+        m_run.return_value = cp
+        expected_result = [multipath.MMap(*mp_out.split())._asdict()]
+        result = multipath.multipath_show_maps()
+        self.assertEqual(expected_result, result)
+
+    @mock.patch('probert.multipath.subprocess.run')
+    def test_multipath_show_maps_skips_unparsable_output(self, m_run):
```
Comment (OddBloke) [inline probert/tests/test_multipath.py]: Same question/comment as https://github.com/CanonicalLtd/probert/pull/62/files#r277758404

File: probert/multipath.py (hunk 1 above)
Comment (raharper) [reply probert/multipath.py]: Yes

File: probert/tests/helpers.py (hunk 2 above)
Comment (raharper) [reply probert/tests/helpers.py]: It can, I'll move it.

File: probert/tests/test_multipath.py (hunk 3 above)
Comment (raharper) [reply probert/tests/test_multipath.py]: I'll fix.  Looks like we should add pep8 to make lint target

File: probert/tests/test_multipath.py (hunk 4 above)
Comment (raharper) [reply probert/tests/test_multipath.py]: Yes, we're bionic+ only

File: probert/tests/test_multipath.py (hunk 5 above)
Comment (raharper) [reply probert/tests/test_multipath.py]: The input we parse here is from the multipathd show <paths|maps> with a formatter input.  In the case that multipathd doesn't have a value, sometimes it outputs "".  That makes me think that we should update the format to include the field names so we know which one was missing.  I'll do that in a separate PR and then update this unittest.

File: probert/tests/test_multipath.py (hunk 6 above)
Comment (raharper) [reply probert/tests/test_multipath.py]: Same answer.

File: probert/tests/test_multipath.py (hunk 3 above)
Comment (OddBloke) [reply probert/tests/test_multipath.py]: I've filed #63 for that

File: probert/multipath.py (hunk 1 above)
Comment (OddBloke) [reply probert/multipath.py]: This change ended up on the wrong line, I think.

Comment (raharper) [reply probert/multipath.py]: Oh, I see. Yes

**Comment by powersj** [comment]
@OddBloke is this ready for merge? Or is there still an outstanding comment?
